ai/producer.go

```diff
@@ -171,12 +171,24 @@ func (ai *aiPlayer) buildOrUpgradeStarbase(planet *cs.Planet) error {
 	// if we are being targeted for bombing though, we want to try and build a starbase regardless
 	// TODO: Add ability to build fuel depots and infrastructure based on a (lower) cutoff
 	// This will be useful for IT/PP and desperately necessary for AR
-	planetaryStructuresBuilt := min(float64(planet.Mines)/float64(planet.Spec.MaxMines), float64(planet.Factories)/float64(planet.Spec.MaxFactories))
-	if !(targeted || attackShipsInOrbit) && planetaryStructuresBuilt < ai.config.fleetProductionCutoff {
-		// this will need to be changed for -f/AR races to work as
-		// they don't build mines & such regardless
-		// AR in particular will require entirely separate logic
-		return nil
+	if !(targeted || attackShipsInOrbit) {
+		if ai.Player.Race.Spec.InnateResources {
+			if !planet.Spec.CanTerraform { // Terraforming is economic development for AR planets
+				existingDesign := ai.GetDesign(planet.Spec.StarbaseDesignNum)
+				if len(existingDesign.Slots) == 0 && existingDesign != ai.fuelDepotDesign {
+					// Bigger starbase allows bigger population
+					ai.addStarbaseToTopOfQueue(planet, ai.fuelDepotDesign)
+				} else {
+					ai.upgradeStarbase(planet, ai.config.minYearsToQueueStarbasePeaceTime)
+				}
+			}
+			return nil
```

Comment:
Can you make this an early return instead of a nested if?
```suggestion
			if planet.Spec.CanTerraform {
			        // Terraforming is economic development for AR planets
			        return nil
			}
		        existingDesign := ai.GetDesign(planet.Spec.StarbaseDesignNum)
			if len(existingDesign.Slots) == 0 && existingDesign != ai.fuelDepotDesign {
				// Bigger starbase allows bigger population
				ai.addStarbaseToTopOfQueue(planet, ai.fuelDepotDesign)
			} else {
				ai.upgradeStarbase(planet, ai.config.minYearsToQueueStarbasePeaceTime)
			}
```

Reply:
One thing I'm not sure about, what happens when you have a starter world. Won't it always immediately queue up a larger starbase even if the planet isn't ready for it?

Reply:
I mean if the planet can't terraform anymore it's not as if it can do much else on a starter colony lol.
(Probably need research allocation yeah but that's a later issue lmao)

Reply:
The nested if has been replaced by an early return.

Regarding planets not being ready for larger star bases: As Bertie suggests, a starter colony that has finished its terraforming can only contribute to research or build a star base. The current code already directs the AI to use the largest star base hull available, even if it is only as a fuel depot. A fuel depot will enable faster colonist transfer and return flights. Also, having already built the same star base hull as the one that will be built for protection (since both use the largest available) means that the protection star base will be quicker to build (because cheaper) if the hull does not need to be upgraded when the planet becomes ```targeted```.

Also, an Ultra station costs the same Ironium, Boranium and resources as a Space Station, the only difference is 50 Germanium. A Death Star costs the same as a Space Station except for 100 Germanium and 150 resources. It seems better to pay these slightly higher initial prices in the first place than to pay the (50%) upgrade costs later.

However, a Space Dock is substantially cheaper than a Space Station. When playing AR, I normally upgrade to a Space Dock immediately after colonizing to get the refueling, even before I have started terraforming. However, the current Star Base design code has no circumstances that will result in the use of a Space Dock hull in a Star Base design, so enabling use of Space Docks for worlds below a certain population would require a much bigger patch than the current one, including a whole new ShipDesignPurpose.

By the way: Having the AI upgrade to a Space Dock before finishing terraforming turns out to be a bad idea. I have recently been playing against several AR AIs. In order to give them a boost I changed their Starter Colonies to use the Space Dock hull instead of the Orbital Fort hull, to simulate a human player upgrading immediately after colonizing. Your question about not being ready for a larger star base encouraged me to log in as one of the AIs to see what the AI have been doing. It turns out that having a Space Dock before finishing their terraforming led to them putting several years-worth of ships in their build queues. All their planets are suffering an iron crunch, so their queues are blocked, and they don't get to the terraforming at the bottom of the queue. AR not terraforming is just as bad as a non-AR not building factories. Also, since I am already using the patch being submitted, this had the side effect of them not upgrading star bases, even for planets that had quite high populations and would normally have completed their terraforming by now. I had made some of their nearby planets to be already perfect for them as another boost. These did get upgraded star bases.